src/Symfony/Component/HttpClient/CachingHttpClient.php

```diff
@@ -11,62 +11,346 @@
-        }
+        if ($store instanceof StoreInterface) {
+            trigger_deprecation('symfony/http-client', '7.3', 'Passing a "%s" as constructor\'s 2nd argument of "%s" is deprecated, "%s" expected.', StoreInterface::class, __CLASS__, TagAwareAdapterInterface::class);
 
-        $kernel = new HttpClientKernel($client);
-        $this->cache = new HttpCache($kernel, $store, null, $defaultOptions);
+            if (!class_exists(HttpClientKernel::class)) {
+                throw new \LogicException(\sprintf('Using "%s" requires the HttpKernel component, try running "composer require symfony/http-kernel".', __CLASS__));
+            }
+
+            $kernel = new HttpClientKernel($client);
+            $this->cache = new HttpCache($kernel, $store, null, $defaultOptions);
 
-        unset($defaultOptions['debug']);
-        unset($defaultOptions['default_ttl']);
-        unset($defaultOptions['private_headers']);
-        unset($defaultOptions['skip_response_headers']);
-        unset($defaultOptions['allow_reload']);
-        unset($defaultOptions['allow_revalidate']);
-        unset($defaultOptions['stale_while_revalidate']);
-        unset($defaultOptions['stale_if_error']);
-        unset($defaultOptions['trace_level']);
-        unset($defaultOptions['trace_header']);
+            unset($defaultOptions['debug']);
+            unset($defaultOptions['default_ttl']);
+            unset($defaultOptions['private_headers']);
+            unset($defaultOptions['skip_response_headers']);
+            unset($defaultOptions['allow_reload']);
+            unset($defaultOptions['allow_revalidate']);
+            unset($defaultOptions['stale_while_revalidate']);
+            unset($defaultOptions['stale_if_error']);
+            unset($defaultOptions['trace_level']);
+            unset($defaultOptions['trace_header']);
+        } else {
+            $this->cache = $store;
+        }
 
         if ($defaultOptions) {
             [, $this->defaultOptions] = self::prepareRequest(null, null, $defaultOptions, $this->defaultOptions);
         }
     }
 
     public function request(string $method, string $url, array $options = []): ResponseInterface
+    {
+        if ($this->cache instanceof HttpCache) {
+            return $this->legacyRequest($method, $url, $options);
+        }
+
+        [$fullUrl, $options] = self::prepareRequest($method, $url, $options, $this->defaultOptions);
+
+        $fullUrl = implode('', $fullUrl);
+        $fullUrlTag = hash('xxh3', $fullUrl);
+
+        if (\in_array($method, self::UNSAFE_METHODS, true)) {
+            $this->cache->invalidateTags([$fullUrlTag]);
+        }
+
+        if (!empty($options['body']) || isset($options['normalized_headers']['range']) || !\in_array($method, self::CACHEABLE_METHODS, true)) {
+            return $this->client->request($method, $url, $options);
```

Comment:
you are still directly returning the inner response here in some cases, while the `stream` method assumes that any AsyncResponse is owned by this CachingHttpClient, which will break things if the inner client is also an async decorator.
and this might also cause issues if the inner client is a MockClient as `stream` also takes ownership of the mock responses.

You might need to create a pass-through response instead, so that you also return a response owned by this client.

Reply:
adressed in https://github.com/symfony/symfony/pull/59576/commits/33fb4eaed827cb9521333d165c1399668826807a